File: app/src/main/java/jp/gcreate/product/filteredhatebu/CustomApplication.java
package jp.gcreate.product.filteredhatebu;

import android.app.Activity;
import android.app.Application;
import android.content.Context;

import com.jakewharton.threetenabp.AndroidThreeTen;
import com.squareup.picasso.Picasso;

import java.util.HashMap;

import javax.inject.Inject;

import jp.gcreate.product.filteredhatebu.di.ActivityComponent;
import jp.gcreate.product.filteredhatebu.di.ActivityModule;
import jp.gcreate.product.filteredhatebu.di.AppComponent;
import jp.gcreate.product.filteredhatebu.di.AppModule;
import jp.gcreate.product.filteredhatebu.di.DaggerAppComponent;
import jp.gcreate.product.filteredhatebu.di.qualifier.ApplicationContext;
import jp.gcreate.product.filteredhatebu.util.CrashlyticsWrapper;
import jp.gcreate.product.filteredhatebu.util.StethoWrapper;
import timber.log.Timber;

/**
 * Copyright 2016 G-CREATE
 */

public class CustomApplication extends Application {
    private AppComponent appComponent;
    private HashMap<String, ActivityComponent> activityComponentMap = new HashMap<>();
    private HashMap<String, ActivityModule>    activityModuleMap    = new HashMap<>();
    @Inject @ApplicationContext
    Context      context;
    @Inject
    StethoWrapper stetho;
    @Inject
    Timber.Tree tree;
    @Inject
    Picasso.Builder picassoBuilder;
    @Inject
    CrashlyticsWrapper crashlyticsWrapper;

    public static AppComponent getAppComponent(Context context) {
        CustomApplication application = (CustomApplication) context.getApplicationContext();
        return application.appComponent;
    }

    public static ActivityComponent getActivityComponent(Activity activity) {
        String key = activity.getClass().getSimpleName();
        Timber.d("getActivityComponent: activity name:%s", key);
        CustomApplication application = (CustomApplication) activity.getApplication();
        if (application.hasComponent(key)) {
            application.getActivityModule(key).updateContext(activity);
            return application.getActivityComponent(key);
        } else {
            AppComponent appComponent = getAppComponent(activity);
            ActivityModule module = new ActivityModule(activity);
            ActivityComponent activityComponent = appComponent.plus(module);
            application.setActivityComponent(key, activityComponent);
            application.setActivityModule(key, module);
            return activityComponent;
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();

        appComponent = DaggerAppComponent.builder()
                                         .appModule(new AppModule(this))
                                         .build();
        appComponent.inject(this);

        Timber.plant(tree);
        stetho.install();
        crashlyticsWrapper.init(this);
        AndroidThreeTen.init(this);
        Picasso.setSingletonInstance(picassoBuilder.build());

        Timber.d("application:%s", context.toString());
    }

    public boolean hasComponent(String key) {
        return activityComponentMap.containsKey(key) && activityModuleMap.containsKey(key);
    }

    public ActivityComponent getActivityComponent(String key) {
        return activityComponentMap.get(key);
    }

    public void setActivityComponent(String key, ActivityComponent component) {
        activityComponentMap.put(key, component);
    }

    public void setActivityModule(String key, ActivityModule module) {
        activityModuleMap.put(key, module);
    }

    public ActivityModule getActivityModule(String key) {
        return activityModuleMap.get(key);
    }
}

File: app/src/main/java/jp/gcreate/product/filteredhatebu/CustomApplication.kt
package jp.gcreate.product.filteredhatebu

import android.app.Application
import androidx.work.*
import com.jakewharton.threetenabp.AndroidThreeTen
import com.squareup.picasso.Picasso
import jp.gcreate.product.filteredhatebu.di.*
import jp.gcreate.product.filteredhatebu.domain.CrawlFeedsWork
import jp.gcreate.product.filteredhatebu.domain.DeleteFeedsWork
import jp.gcreate.product.filteredhatebu.ui.common.NotificationUtil
import org.koin.android.ext.android.getKoin
import org.koin.android.ext.android.inject
import org.koin.android.ext.koin.androidContext
import org.koin.androidx.workmanager.koin.workManagerFactory
import org.koin.core.context.startKoin
import timber.log.Timber
import java.util.concurrent.TimeUnit

/**
 * Copyright 2016 G-CREATE
 */
class CustomApplication : Application(), Configuration.Provider {

    private val tree: Timber.Tree by inject()
    private val picassoBuilder: Picasso.Builder by inject()
    private val notificationUtil: NotificationUtil by inject()

    override fun onCreate() {
        super.onCreate()

        startKoin {
            androidContext(this@CustomApplication)
            modules(
                listOf(
                    koinAppModule,
                    koinDataModule,
                    koinDebugModule,
                    koinNetworkModule,
                    koinViewModelModule
                )
            )
            workManagerFactory()
        }

        Timber.plant(tree)
        AndroidThreeTen.init(this)
        Picasso.setSingletonInstance(picassoBuilder.build())
        scheduleCrawlFeedWork()
        DeleteFeedsWork.schedule()
        notificationUtil.installChannel()
    }

    override fun getWorkManagerConfiguration(): Configuration {
        val workerFactory = getKoin().getAll<WorkerFactory>()
        val config = Configuration.Builder()
        workerFactory.forEach { config.setWorkerFactory(it) }
        return config.build()
    }

    private fun scheduleCrawlFeedWork() {
        val constraints = Constraints.Builder().apply {
            setRequiredNetworkType(NetworkType.UNMETERED)
            setRequiresBatteryNotLow(true)
            setRequiresDeviceIdle(true)
        }
            .build()
        val request = PeriodicWorkRequestBuilder<CrawlFeedsWork>(6, TimeUnit.HOURS)
            .setConstraints(constraints)
            .addTag("repeat_crawling")
            .setInputData(workDataOf(CrawlFeedsWork.KEY_TYPE to "period"))
            .build()
        WorkManager.getInstance(this)
            .enqueueUniquePeriodicWork("repeat_crawling", ExistingPeriodicWorkPolicy.KEEP, request)
    }
}
